includes/class-wc-calypso-bridge-introductory-offers.php

```diff
@@ -0,0 +1,84 @@
+<?php
+/**
+ * Contains the logic for querying plans with introductory offers.
+ *
+ * @package WC_Calypso_Bridge/Classes
+ * @since   1.0.6
+ * @version 2.0.8
+ */
+
+use Automattic\Jetpack\Connection\Client;
+use Automattic\Jetpack\Connection\Manager;
+
+/**
+ * WC_Calypso_Bridge_Introductory_offers
+ */
+class WC_Calypso_Bridge_Introductory_offers {
+
+	CONST TRANSIENT_PREFIX = 'wc-calypso-bridge-introductory-plans-';
+
+	/**
+	 * return introductory offers for the current blog.
+	 *
+	 * @return array|mixed
+	 */
+	public static function get_introductory_offers_for_current_blog() {
+		return static::get_introductory_offers( Jetpack_Options::get_option( 'id' ) );
+	}
+
+	/**
+	 * Format offer data for JS.
+	 *
+	 * @param $offer
+	 *
+	 * @return array
+	 */
+	public static function extract_offer_data_for_js( $offer ) {
+		return array(
+			'rawPrice' => $offer['introductory_offer_raw_price'],
+			'formattedPrice' => $offer['introductory_offer_formatted_price'],
+			'intervalUnit' => $offer['introductory_offer_interval_unit'],
+			'intervalCount' => $offer['introductory_offer_interval_count'],
+			'formattedIntervalUnit' => $offer['introductory_offer_interval_count'] === 1 ? $offer['introductory_offer_interval_unit'] : $offer['introductory_offer_interval_unit'].'s',
+		);
+	}
+
+	/**
+	 * Return introductory offers by blog I.D
+	 *
+	 * @param $blog_id
+	 *
+	 * @return array|mixed
+	 */
+	public static function get_introductory_offers( $blog_id ) {
+		if ( ! ( new Manager() )->is_user_connected() ) {
+			return [];
+		}
+
+		$cached_offers = get_transient( static::TRANSIENT_PREFIX . $blog_id );
+		if ( $cached_offers ) {
+			return $cached_offers;
+		}
+
+		$response = Client::wpcom_json_api_request_as_blog(
+			'/sites/'.$blog_id.'/introductory-offers',
+			'1.3',
+			array(),
+			null,
+			'rest'
+		);
+
+		$offers = [];
+
+		if ( ! is_wp_error( $response ) && isset( $response[ 'http_response' ] ) && $response[ 'http_response' ] instanceof WP_HTTP_Requests_Response && 200 === $response[ 'http_response' ]->get_status() ) {
+			$data = json_decode( $response['http_response']->get_data(), true );
+			if ( is_array( $data ) && count( $data ) ) {
+				 $offers = $data;
+			}
+		}
+
+		set_transient( static::TRANSIENT_PREFIX . $blog_id, $offers, 120 );
```

Comment:
Good thinking to set transient for performance!

There 2 things that may impact our cache:

1. Current blog's plan change
2. Currency change (since we display the price in UI)

**Plan**: What if we also append the current plan slug to the transient name? Though it's probably not easy to determine the exact current plan slug ([example](https://github.com/Automattic/wc-calypso-bridge/blob/78bf2dcc2cb3f625428cf8315ae24e1dfeeb2472/class-wc-calypso-bridge-dotcom-features.php#L298) where we detect them via purchased features). Perhaps we can serialize all purchases and make that as a key.

**Currency**: I'm not aware of any way to get the latest user currency, but it should be somewhere.

But I digress, I don't think this is a blocker since if there's any impact, it'll probably be edge case.

Reply:
That's a good approach, but that means we need to call wpcom_get_site_purchases() on every check unless I'm mistaken. I figured 2 mins is short enough for the page to be refreshed when the user decided to upgrade.

1. User sees Woo Home with cached data. 2 mins cache starts.
2. User clicks `Upgrade` and goes through the flow.
3. User comes back to Woo Home. Cache should have been expired at this point.

Reply:
> That's a good approach, but it implies that we need to call wpcom_get_site_purchases() with every check, unless I'm mistaken

I think `wpcom_get_site_purchases` is already executed in every request anyway since there are fundamental logic built around it.

Still, good point on 2 minutes buffer for upgrades, I think it's a reasonable. 👍

Reply:
Sounds good. Let's skip it for now 👍